### PR TITLE
ReWorks™

### PR DESCRIPTION
**Pull request by Reboot-Codes**
# ReWorks™
This is a large rework of a lot of todo's that have been bugging me. First is the [site](https://discordeno.mod.land), which just leaves prospective developers to figure it out by themselves. Next is the 4 templates we have. We only need 2 or 3 (Big-Bot, Starter, and maybe Node.js). Third is the `fileloader` plugin; in short, we shouldn't be writing anything to disk. Last is the lack of a deno task to properly format PR's.

> Note: Big Bot Guide is on hold until v13 is ready.

## General:
- [x] Add `deno task fmt` to make sure PR's are properly formatted.
- [ ] Update all `README.md`s

## Site:
- [x] Working on starter guide, and syncing with minimal example.
- [ ] Working on big-bot guide, and syncing with big-bot example.
- [ ] Update Site Homepage (and maybe theme with blurple)

## Templates:
- [ ] Update other templates to work with RC 45 (as is the mostly, stable version).

## Plugins
- [ ] Fix file loader (we shouldn't be writing anything to disk).

### CODE REVIEW

**Comment by Reboot-Codes** [comment]
All of the changes to the `/template` directory were by Its200550705, my changes were to `/site`, `.vscode/` and `deno.json`. Git is really weird. 

"Reformat Site Documentation Flow" is a reordering of the sidebar and some wording changes.

File: site/docs/starter/project-structure.md [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+  // Start the bot
+  await startBot(bot);
+} else {
+  throw "Error: No Token!";
+}
+```
+
+## `src/utils/logger.ts`
+
+We should have a nicer way to log information, so we will use this logger (you can find it in most, if not all of our
```
Comment (Skillz4Killz) [inline site/docs/starter/project-structure.md:134]: If we really do use it that much, thoughts on just moving this code to another repo like discordeno/logger and having a discordeno_logger dep on deno.land

File: site/docs/starter/slash-commands.md [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+
+if (BOT_TOKEN) {
+  // Start the bot
+  await startBot(bot);
+} else {
+  throw "Error: No Token!";
+}
+```
+
+This then exposes `bot.guilds`, `bot.members`, and more. The sweepers will clean the cache to make sure we aren't using
```
Comment (Skillz4Killz) [inline site/docs/starter/slash-commands.md:64]: I believe this is a big feature in our lib, unlike others. Sweepers' goal is to prevent a memory leak that will crash ur process if u keep caching everything and never clean it out ur process will eventually run out of memory.

File: site/docs/starter/slash-commands.md [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+
+All of our commands will follow the same structure:
+
+```typescript title="src/commands/cmd/ping.ts"
+import { addCommand } from "../mod.ts";
+import { ApplicationCommandTypes } from "../../../deps.ts";
+import { snowflakeToTimestamp, humanizeMilliseconds } from "../../helpers.ts"
+
+export const ping = addCommand({
+  // this command will be used `/ping` in discord.
```
Comment (Skillz4Killz) [inline site/docs/starter/slash-commands.md:123]: I would recommend these comment explanations be done separately below so you can attach images/screenshots of what they look like to the user

File: site/docs/starter/slash-commands.md [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+    );
+  }
+});
+```
+
+We should also add the following helper functions to `src/utils/helpers.ts`:
+
+```typescript title="src/utils/helpers.ts"
+/** Turn a discord snowflake into a millisecond value. */
+export function snowflakeToTimestamp(id: bigint) {
```
Comment (Skillz4Killz) [inline site/docs/starter/slash-commands.md:154]: This should be something in the helper plugin, id recommend using this chance to encourage using that

File: site/docs/starter/slash-commands.md [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+  return Number(id / 4194304n + 1420070400000n);
+}
+
+/** Turn a millisecond value into a human readable string. */
+export function humanizeMilliseconds(milliseconds: number) {
```
Comment (Skillz4Killz) [inline site/docs/starter/slash-commands.md:159]: Once again could be moved to a helper plugin

File: site/docs/starter/slash-commands.md [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+  const minuteString = minutes ? `${minutes}m ` : "";
+  const secondString = seconds ? `${seconds}s ` : "";
+
+  return `${dayString}${hourString}${minuteString}${secondString}`.trim();
+}
+```
+
+## Registering Slash Commands
+
+You may have noticed the `Command.scope` option when defining a command. This is because discord provides two scopes
```
Comment (Skillz4Killz) [inline site/docs/starter/slash-commands.md:180]: I think this will be changing and `bot` scope will be enough

File: site/docs/starter/slash-commands.md [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+        name: command.name,
+        description: command.description,
+        type: command.type,
+        options: command.options ? command.options : undefined,
+      });
+    }
+  }
+
+  if (globalCommands.length && (scope === "Global" || scope === undefined)) {
+    // Due to the way discord internally handles global commands, it might take up to an hour before they are available.
```
Comment (Skillz4Killz) [inline site/docs/starter/slash-commands.md:246]: This is no longer true, they are instant

File: site/docs/starter/slash-commands.md [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+
+  if (perGuildCommands.length) {
+    // Upsert will add commands, replace existing commands, and delete ones that do not exist anymore.
+    await upsertApplicationCommands(bot, perGuildCommands, guild.id);
+  }
+}
+// ...
+```
+
+We'll add this to our `mod.ts` to make sure that we always have the latest commands ready for users:
```
Comment (Skillz4Killz) [inline site/docs/starter/slash-commands.md:291]: I would recommend rephrasing this so that its done on mod.ts so that when u make any changes it will automatically make changes once u restart ur bot.

File: site/docs/starter/slash-commands.md [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+}));
+
+enableCacheSweepers(bot);
+
+if (BOT_TOKEN) {
+  // Start the bot
+  await startBot(bot);
+  // highlight-start
+  // Update all commands
+  await updateCommands(bot);
```
Comment (Skillz4Killz) [inline site/docs/starter/slash-commands.md:332]: I would recommend putting these two in a Promise.all() so it is slightly faster, commands can be updated without having the bot online.

File: site/docs/starter/slash-commands.md [hunk 10]
```diff
@@ -0,0 +1,550 @@
...
+  // highlight-end
+} else {
+  throw "Error: No Token!";
+}
+```
+
+Then add the following to the `guildCreate` event handler to make sure new guilds also get our guild only commands:
```
Comment (Skillz4Killz) [inline site/docs/starter/slash-commands.md:339]: the following? but nothing is shown after this line

File: site/docs/starter/slash-commands.md [hunk 11]
```diff
@@ -0,0 +1,550 @@
...
+  yellow,
+} from "../../deps.ts";
+import { Command, commands } from "../commands/mod.ts";
+import { getGuildFromId, isSubCommand, isSubCommandGroup } from "../utils/helpers.ts";
+import { logger } from "../utils/logger.ts";
+import { events } from "./mod.ts";
+
+const log = logger({ name: "Event: interactionCreate" });
+
+events.interactionCreate = async (rawBot, interaction) => {
```
Comment (Skillz4Killz) [inline site/docs/starter/slash-commands.md:367]: This is quite complicated for a starter guide, what if this was a plugin so that you could do `events.command` `events.button` or `events.modal` etc... similar to events.messageCreate

File: site/docs/starter/slash-commands.md [hunk 12]
```diff
@@ -0,0 +1,550 @@
...
+```
+
+That might seem scary, however, we've included a lot of comments to help you understand what it does. We also need to
+add some more helpers to the `src/utils/helpers.ts` file (make sure to import types from `src/commands/mod.ts` and
+`deps.ts`):
+
+```typescript title="src/utils/helpers.ts"
+// ...
+
+export async function getGuildFromId(
```
Comment (Skillz4Killz) [inline site/docs/starter/slash-commands.md:493]: Why is this needed?

**Review by Skillz4Killz — APPROVED** [review]

File: site/docs/starter/getting-started.md [hunk 13]
```diff
@@ -0,0 +1,28 @@
+# Getting Started
+
+:::caution
+
+This guide is being updated to better explain what discordeno does, how to properly write a bot, and use the latest v13
```
[the remark this answers is not in the document]
Comment (Reboot-Codes) [reply site/docs/starter/getting-started.md:5]: This was in case I forgot to finish this guide before PR, it will be removed on pr merge.

File: site/docs/starter/event-handlers.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+and messages in channels the bot has access to.
+
+:::note Guilds vs Servers
+
+The term "Guild" is the technical name for a "Server". This can be confusing sometimes. But a group of channels on the
+discord client which is called a "Server" is actually a "Guild" on the backend.
+
+:::
+
+## `messageCreate` Handler
```
[the remark this answers is not in the document]
Comment (Reboot-Codes) [reply site/docs/starter/event-handlers.md:71]: Defining handlers in `mod.ts` is discouraged. Later pages define event handlers in a new directory.

File: site/docs/starter/getting-started.md [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+version to ensure stability, consistency and the best developer experience. This guide serves as the purpose for
+introducing Discordeno's (very) opinionated method of writing bots to new developers.
+
+This guide covers creating a very simple bot meant for just a few servers. If you would like to just clone the code, we
+have a [minimal template](https://github.com/discordeno/discordeno/blob/main/template/minimal) ready to go, but make
+sure to follow the [setup instructions](https://github.com/discordeno/discordeno/tree/main/template/minimal#readme).
+
+## Requirements
+
+- The latest version of [Deno](https://deno.land/).
```
[the remark this answers is not in the document]
Comment (Reboot-Codes) [reply site/docs/starter/getting-started.md:20]: Guides will be cross platform, working on it...

File: site/docs/starter/slash-commands.md (hunk 11 above)
Comment (Reboot-Codes) [reply site/docs/starter/slash-commands.md:367]: Like in a DD plugin?

File: site/docs/starter/slash-commands.md (hunk 12 above)
Comment (Reboot-Codes) [reply site/docs/starter/slash-commands.md:493]: As in the `// ...`? That would signify earlier file contents. As in why `getGuildFromId()` is needed? I remember that the bot's helper for this won't automatically add the guild to the cache, therefore we use this function instead.

File: site/docs/starter/slash-commands.md (hunk 2 above)
Comment (Reboot-Codes) [reply site/docs/starter/slash-commands.md:64]: Should I add this reason?

File: site/docs/starter/slash-commands.md (hunk 3 above)
Comment (Reboot-Codes) [reply site/docs/starter/slash-commands.md:123]: Good Idea.

File: site/docs/starter/slash-commands.md (hunk 8 above)
Comment (Reboot-Codes) [reply site/docs/starter/slash-commands.md:291]: I'll update the guide to use `deno run --watch`

**Comment by deepsarda** [comment]
Perhaps use husky, to automatically format code on commit?

**Comment by lts20050703** [comment]
@Skillz4Killz @itohatweb what are your thoughts?

**Comment by lts20050703** [comment]
@deepsarda image

**Comment by lts20050703** [comment]
@Reboot-Codes please resolve the merge conflicts
image

**Comment by Reboot-Codes** [comment]
On it.

**Comment by Skillz4Killz** [comment]
@Reboot-Codes any news on thsi pr, its been open a while and keeps getting more and more conflicts? id love to merge but i feel its incomplete so i hesitate every time. Maybe try and split them into smaller prs?

**Comment by Reboot-Codes** [comment]
Hi Skillz, yes, this PR is quite old and incomplete. I'll probably re-make it with the content of the pages and the current Lib code. (Been a bit busy with school, I apologize for that.)

**Comment by Skillz4Killz** [comment]
Take your time, no need to apologize!
